[V1][VLM] Proper memory profiling for image language models

ywang96 opened the pull request:
This PR adds memory profiling for image language models in V1 so that peak memory usage is correctly measured to avoid potential CUDA OOM.

A few things to note:
1. We still use the dummy `multi_modal_data` for memory profiling. This not only allows us to reuse the existing code, but is also developer-friendly since defining a dummy PIL image is much more intuitive than defining dummy pixel values. In order to do so, `mm_input_mapper` is added to model runner only for profiling purpose.

2. This current version only works for image (because of limitation on `mm_input_mapper`). The memory profiling section has supported profiling for other single non-text modality models, but may require some additional design to be extended to mixed non-text modality profiling.

3. Encoder compute budget and cache size are added to scheduler config, but they remain hardcoded and should be optimized in a later PR.

Verified with `VLLM_USE_V1=1 VLLM_ENABLE_V1_MULTIPROCESSING=1 python3 mmmu_bench.py --model OpenGVLab/InternVL2_5-8B --trust-remote-code --gpu-memory-utilization 0.99` from #11196 

Main branch:
```
...
INFO 12-15 10:18:11 uniproc_executor.py:64] # GPU blocks: 31207
...
torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 322.00 MiB. GPU 0 has a total capacity of 79.10 GiB of which 249.88 MiB is free. 
```

This PR:
```
...
INFO 12-15 10:16:26 uniproc_executor.py:64] # GPU blocks: 30973
...
Request throughput: 8.13 req/s
Total generated tokens: 39100
Token generation rate: 317.89 tok/s
```

ywang96 committed:
add

Signed-off-by: Roger Wang <ywang@roblox.com>

ywang96 committed:
comment

Signed-off-by: Roger Wang <ywang@roblox.com>

ywang96 committed:
iterate

Signed-off-by: Roger Wang <ywang@roblox.com>

ywang96 committed:
comment

Signed-off-by: Roger Wang <ywang@roblox.com>

DarkLight1337 commented on vllm/multimodal/registry.py:
```diff
@@ -200,6 +200,23 @@ def register_max_image_tokens(
         """
         return self.register_max_multimodal_tokens("image", max_mm_tokens)
 
+    def get_max_tokens_per_item_by_modality(
+        self,
+        model_config: "ModelConfig",
+    ) -> Mapping[str, int]:
+        """
+        Get the maximum number of tokens per data item from each modality
+        for profiling the memory usage of a model.
+
+        Note:
+            This is currently only used in V1.
+        """
+
+        return {
+            key: plugin.get_max_multimodal_tokens(model_config)
+            for key, plugin in self._plugins.items()
+        }
+
```
Can we update `get_max_tokens_by_modality` to be based on this new method?

DarkLight1337 commented on vllm/v1/core/scheduler.py:
```diff
@@ -73,14 +73,13 @@ def __init__(
-        # FIXME(woosuk): Below are placeholder values. We need to calculate the
-        # actual values from the configurations.
-        self.max_num_encoder_input_tokens = 16384
+        self.max_num_encoder_input_tokens = self.scheduler_config.max_num_encoder_input_tokens  #noqa: E501
```
```suggestion
        self.max_num_encoder_input_tokens = self.scheduler_config.max_num_encoder_input_tokens  # noqa: E501
```

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -604,6 +607,52 @@ def profile_run(self) -> None:
+            # a single request, therefore here we always replicate first
+            # item by max_num_mm_items times.
+            batched_dummy_mm_inputs = MultiModalKwargs.batch(
+                [dummy_mm_kwargs[0] for _ in range(max_num_mm_items)])
```
```suggestion
                [dummy_mm_kwargs[0]] * max_num_mm_items)
```

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -604,6 +607,52 @@ def profile_run(self) -> None:
+                **batched_dummy_mm_inputs)
+
+            # Cache the dummy encoder outputs.
+            dummy_req_input_ids = [("0", i) for i in range(max_num_mm_items)]
+            self.encoder_cache["0"] = {}
+            for (req_id, input_id), output in zip(dummy_req_input_ids,
+                                                  dummy_encoder_outputs):
+                self.encoder_cache[req_id][input_id] = output
```
```suggestion
            dummy_cache_item = {}
            for input_id, output in zip(range(max_num_mm_items),
                                        dummy_encoder_outputs):
                dummy_cache_item[input_id] = output
            self.encoder_cache["0"] = dummy_cache_item
```
Simplify the code a bit

ywang96 replied:
That's a good point!

ywang96 committed:
address comments

Signed-off-by: Roger Wang <ywang@roblox.com>

DarkLight1337 reviewed (approved):
LGTM now!

Commit:
fix empty tensor

Commit:
fix empty tensor

Signed-off-by: Roger Wang <ywang@roblox.com>

ywang96 committed:
Merge branch 'mm_profiling' of https://github.com/ywang96/vllm into mm_profiling

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -604,6 +607,54 @@ def profile_run(self) -> None:
+            # Create dummy batch of multimodal inputs.
+            dummy_mm_data = self.input_registry \
+                .dummy_data_for_profiling(self.model_config,
+                                            self.max_num_tokens,
+                                            self.mm_registry).multi_modal_data
```
nit: Can you please re-format this? 😅

ywang96 replied:
This is auto formatted by our `format.sh` script, but I can modify the code here

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -604,6 +607,54 @@ def profile_run(self) -> None:
+                self.mm_registry.get_max_tokens_per_item_by_modality(
+                    self.model_config).values())
+
+            max_num_mm_items = min(
+                self.max_num_encoder_input_tokens,
+                self.encoder_cache_size) // max_tokens_per_mm_item
```
A dumb question: in our current implementation, shouldn't `encoder_cache_size` be always larger than or equal to `max_num_encoder_input_tokens`?

mgoin replied:
Why should the encoder cache size affect the maximum tokens we can process as input? It seems we should make the decision if we can allow the cache size to be smaller than the max number of encoder input tokens

WoosukKwon replied:
@ywang96 @mgoin IIUC, we always stores the encoder outputs to the encoder cache, and then use them in the decoder. Therefore, if the encoder cache is smaller than that of the encoder outputs (which is in turn equal to the number of the encoder inputs).

ywang96 replied:
Since both values are hardcoded to 16384 anyways, I will merge this PR as is and revisit this particular logic later when we tune the space & compute budget

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -604,6 +607,54 @@ def profile_run(self) -> None:
+                self.max_num_encoder_input_tokens,
+                self.encoder_cache_size) // max_tokens_per_mm_item
+
+            # Dummy data definition in V0 may contain multiple items for
+            # a single request, therefore here we always replicate first
+            # item by max_num_mm_items times.
```
Can you elaborate more? What does "multiple items" mean here? Does it mean multiple images?

ywang96 replied:
Yes, the dummy data in a dummy sequence for an LMM is based on `limit-mm-per-prompt`. This means if the user sets up a engine with `limit-mm-per-prompt image=2`, then the dummy data for a single sequence will contain two identical dummy images. We had to do this because in V0 encoder call is scheduled together with decoder call.

I will modify the comment here to clarify.

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -604,6 +607,54 @@ def profile_run(self) -> None:
+                "due to the 'get_multimodal_embeddings' method of the model "
+                "not implemented correctly.")
+
+            # Cache the dummy encoder outputs.
+            self.encoder_cache["0"] = dict(enumerate(dummy_encoder_outputs))
```
nit: what about "tmp" or something like that instead?

WoosukKwon reviewed (approved):
Thanks for the PR! Left some minor comments.

Commit:
address comments

ywang96 commented:
@WoosukKwon I've addressed your comments via [1720f68](https://github.com/vllm-project/vllm/pull/11210/commits/1720f680901f18880587ae3fb42470c6f1b70ed7) - Let me know if there's any other concern!

mgoin reviewed (approved):
LGTM thanks!
